### PR TITLE
RFC: Update API to enable connection re-use, allow streaming…

### PR DESCRIPTION
**Pull request by pmeinhardt**
Related to: https://github.com/bitcrowd/sshkit.ex/issues/41

### Description

Update the top-level SSHKit API in order to reuse SSH connections across SSHKit operations, i.e. `SSHKit.run/3`, `SSHKit.upload/4` and `SSHKit.download/4`.

The updated API could look as follows (feedback welcome):

```elixir
hosts = ["1.eg.io", {"2.eg.io", port: 2222}]

# UPDATED: The context no longer receives a list of hosts/a single host.
# It exclusively manages the execution context on the remotes.
context =
  SSHKit.context()
  |> SSHKit.path("/var/www/phx")
  |> SSHKit.user("deploy")
  |> SSHKit.group("deploy")
  |> SSHKit.umask("022")
  |> SSHKit.env(%{"NODE_ENV" => "production"})

# NEW: Explicitly establish a connection
{:ok, conn} = SSHKit.connect(hosts)

# UPDATED: The top-level functions now require a conn struct.
{:ok, conn} = SSHKit.upload(conn, context, ".", recursive: true)
{:ok, conn} = SSHKit.run(conn, context, "yarn install")

# IDEA: Passing a context could be optional (different possible versions)
{:ok, conn} = SSHKit.run(conn, "apt-get install -y postgresql", context: context)
{:ok, conn} = conn |> SSHKit.within(context) |> SSHKit.run(conn, "apt-get install -y postgresql")

# NEW: Explicitly close the connection once we're done
{:ok, _} = SSHKit.close(conn)
```

cf. current [README.md](https://github.com/bitcrowd/sshkit.ex/blob/master/README.md#usage)

**Additional considerations**

A major idea/API change that may play into this draft is how we handle stdout/stderr streaming which is an interesting case we did not handle in SSHKit 0.x but is definitely interesting to give quicker user feedback, e.g. on deployment cases as with bootleg (see the [alt. `capture` implementation](https://github.com/labzero/bootleg/blob/5dcbd483c4099359387d2e9a8cb7e93a923422c2/lib/bootleg/ssh.ex#L195-L222)).

This might solve/play into #53 as well.

**Make context optional?**

With the API update, context could be made an optional or keyword argument, because for some things you want to run, you might not need any user/group/path/env/umask changes.

**Update host struct**

There was a request to include host information in result tuples #132 and Bootleg defines a `%Host{}` struct to assign a `name`, we should consider adding  `name` and possibly `tags` to hosts.

### Motivation and Context

At the moment, SSHKit is establishing a new connection for every command. The outlined API changes could allow us to reuse established connections for multiple consecutive operations and improve the speed with which server automation tasks can be carried out. Deployments with [bootleg](https://github.com/labzero/bootleg) (cc @labzero) for instance could benefit from this.

### Discussion: Alternative Designs

As #41 outlines, we could also use a (GenServer) process to cache connections. This however means that it would be harder/not possible to - for instance - open multiple connections to the same server.

It also seems more explicit and conceptually cleaner to separate execution context (`SSHKit.context/0`) from connection information (the new `conn` struct).

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes (with unit and/or functional tests).
- [ ] I have added a note to CHANGELOG.md if necessary (in the `## master` section).

### CODE REVIEW

**Comment by pmeinhardt** [comment]
@brienw, @holetse, @rjanja, @svoynow here's a quick draft on how we could enable connection reuse with SSHKit.ex. Since this could possibly speed up deployments, I thought you might be interested 🙂 

We haven't updated the package in a while and now with some distance I have a few new ideas.

I'd be happy about your thoughts on the API design. It adds another parameter to the main functions, but at the same time cleanly separates (execution) context from connection information.

File: README.md [hunk 1]
```diff
@@ -17,15 +17,19 @@ repeatable way, e.g. in the context of deployment tools:
...
   |> SSHKit.user("deploy")
   |> SSHKit.group("deploy")
   |> SSHKit.umask("022")
   |> SSHKit.env(%{"NODE_ENV" => "production"})
 
-:ok = SSHKit.upload(context, ".", recursive: true)
-:ok = SSHKit.run(context, "yarn install")
+{:ok, conn} = SSHKit.connect(hosts)
+
+{:ok, conn} = SSHKit.upload(conn, context, ".", recursive: true)
```
Comment (klappradla) [inline README.md]: I'm missing the overview if this hinders some current usecases, but in my personal opinion, the API would be cleaner and easier to jump in if both, `upload` and `run` would **not** take a `context`, but have all they they need already in the `conn` that is passed it.

While this means, `connect` needs to be called sometime first in order to get a "connection", I think this pipeline will actually feel super natural for anyone using the library.

Comment (pmeinhardt) [reply README.md]: Awesome @klappradla. Thanks a ton for sharing your thoughts! 💚 

My idea was to explicitly separate `conn` and `context`. My current (and by no means only) way of thinking about these is:

1. You establish an SSH **connection** with a remote (basically `ssh user@remote`).
2. You perform several operations via this connection. Each operation executes within a certain **context** (path, env…). Several operations executed via the same connection may use different contexts.
3. You close the connection.

That's why I came to the conclusion that keeping them separate might be a good idea. If we decide to keep both in a single struct, we'd tie the execution context to a specific connection and vice versa.

Can you elaborate a bit more on why you would like to keep them in a single struct? I would love to look at a few more alternatives. Writing them down probably helps 🙂 

I imagine a combined struct variant would look something like this:

```elixir
hosts = […]

{:ok, conn} =
  hosts
  |> SSHKit.connect()
  |> SSHKit.path(…)
  |> SSHKit.user(…)

{:ok, conn} = SSHKit.run(conn, "mix deps.get")

{:ok, _} = SSHKit.close(conn)
```

The downside of this approach seems to me, that it's not as straight-forward to run multiple commands via the same connection in different contexts (e.g. at different paths…). But maybe I am misreading your comment. 😊 

Happy to hear about your ideas 🙇

Comment (klappradla) [reply README.md]: 💡 thanx for the additional explanation 🙏. I did not understand the purpose of both well enought 🙈 . With my increased knowledge, your idea of separating them actually makes a ton of sense and will probably lead to a cleaner and easier to understand API of the whole library.

The way I understood "reusing" the connection with multiple contexts initially was: the `conn` includes the `context` and one can build a new one (with a new context) with passing in an existing `conn` and without actually establishing a new "connection".
A bit like *"mutating the context of a connection*" 😉

Comment (pmeinhardt) [reply README.md]: @klappradla what you had in mind would also work. I imagine it'd look something like this:

```elixir
{:ok, conn} = SSHKit.connect(hosts)

# Run a command without a specific context.
{:ok, conn, output} = SSHKit.run(conn, "whoami")

# Now we run another command through the same connection but as a different user:
{:ok, conn, output2} =
  conn
  |> SSHKit.user("alter_ego")
  |> SSHKit.run("whoami")

# What should conn be now? Does it remember the new user?
```

The last question is why I was thinking to keep both separate. Maybe we can stick our heads together tomorrow, to see whether I really got your idea 😀

**Comment by pmeinhardt** [comment]
Thanks a lot for the great round today @klappradla, @lwassermann and @andreasknoepfle ❤️ ❤️ ❤️ 

Here's a quick braindump:

* Separating `conn` and `context` as outlined above looked like a clear and understandable option, even if it changes function signature and makes calls longer
*  We discussed chaining/piping in the top-level API
  * piping in combination with `{:ok, conn}` and `{:error, …}` tuples didn't seem like the best option so far
  * `with/1` looked like a more idiomatic fit
  * having more "flow" examples (in the readme) might be nice
* Looking at typical use-cases like [labzdero/bootleg](https://github.com/labzero/bootleg) we found that we need to address stdout/stderr streaming
  * there's at least to options: hook functions and messages to the invoking process + `receive`
  * hooks may have a small drawback if you need to aggregate output but could potentially be easier to use in the top-level API
  * `receive` appears a bit more flexible (either plain hooking or aggregation) and (maybe) more idiomatic?

Cheers 👋 and again much ❤️

**Comment by pmeinhardt** [comment]
As an additional note: [mint](https://github.com/ericmj/mint) uses a message-based API that we can draw inspiration from:

```elixir
{:ok, conn} = Mint.HTTP.connect(:http, "httpbin.org", 80)
{:ok, conn, request_ref} = Mint.HTTP.request(conn, "GET", "/", [], "")

receive do
  message ->
    {:ok, conn, responses} = Mint.HTTP.stream(conn, message)
    IO.inspect responses
end
```

I think this could work very similarly for SSH connect, run, stream…

**Comment by pmeinhardt** [comment]
Closing in favor of #164 😊

**Pull request closed without merging**